In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('videosUS.csv')

In [3]:
df.head()

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   

   category_id              publish_time  \
0         22.0  2017-11-13T17:13:01.000Z   
1         24.0  2017-11-13T07:30:00.000Z   
2         23.0  2017-11-12T19:05:24.000Z   
3         24.0  2017-11-13T11:00:04.000Z   
4         24.0  2017-11-12T18:01:41.000Z   

                                                tags      views     likes  \
0                                    SHANtell martin   748374.0   57527.0   
1  last week tonight trump presidency|"last week ...  2418783.0   97185.0   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434.0  146033.0   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168.0   10172.0   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731.0  132235.0   

   dislikes  comment_count                                  thumbnail_link  \
0    2966.0        15954.0  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1    6146.0        12703.0  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2    5339.0         8181.0  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3     666.0         2146.0  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4    1989.0        17518.0  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   

  comments_disabled ratings_disabled video_error_or_removed  \
0             False            False                  False   
1             False            False                  False   
2             False            False                  False   
3             False            False                  False   
4             False            False                  False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...  
3  Today we find out if Link is a Nickelback amat...  
4  I know it's been a while since we did this sho...

In [4]:
number_of_missing_rows = len(df) - len(df.dropna())
print(number_of_missing_rows)

1036


In [5]:
n_of_missing_values = df.isnull().sum().sum()
print(n_of_missing_values)

7560


In [6]:
cols_n = dict(df.isnull().sum())
print(cols_n)

print(df['publish_time'].isnull())

{'video_id': 0, 'trending_date': 466, 'title': 466, 'channel_title': 466, 'category_id': 466, 'publish_time': 466, 'tags': 466, 'views': 466, 'likes': 466, 'dislikes': 466, 'comment_count': 466, 'thumbnail_link': 466, 'comments_disabled': 466, 'ratings_disabled': 466, 'video_error_or_removed': 466, 'description': 1036}
0        False
1        False
2        False
3        False
4        False
         ...  
41410    False
41411    False
41412    False
41413    False
41414    False
Name: publish_time, Length: 41415, dtype: bool


# Since the number of missing trending dates, titles, channel, titles, ... is the same (466) we could assume that :
- The Missing Values are in the MNAR category (not randomly missing)
- The videos don't exist (in the platform from where the data was taken)

In [7]:
stats = dict([(col, {}) for col in df if (df[col].dtype == 'float64')])

In [8]:
for col in stats: 
    stats[col]['mean'] = df[col].mean()
    stats[col]['median'] = df[col].median()
    stats[col]['q1'] = df[col].quantile(q=0.25)
    stats[col]['q2'] = df[col].quantile(q=0.5)
    stats[col]['q3'] = df[col].quantile(q=.75)
    std_dev = df[col].std()
    stats[col]['zscore'] = [(df[col][i] - stats[col]['mean'])/(std_dev) for i in range(len(df[col]))]




In [9]:
stats

{'category_id': {'mean': 19.97242911914821,
  'median': 24.0,
  'q1': 17.0,
  'q2': 24.0,
  'q3': 25.0,
  'zscore': [0.2679021303989442,
   0.5321613313265069,
   0.40003173086272553,
   0.5321613313265069,
   0.5321613313265069,
   1.060679733181632,
   0.5321613313265069,
   1.060679733181632,
   -2.5068194793404635,
   0.6642909317902882,
   -0.39274587191996235,
   0.5321613313265069,
   -1.3176530751664315,
   0.6642909317902882,
   -0.657005072847525,
   1.060679733181632,
   0.40003173086272553,
   0.9285501327178508,
   1.060679733181632,
   0.9285501327178508,
   0.9285501327178508,
   1.060679733181632,
   0.9285501327178508,
   0.7964205322540695,
   0.40003173086272553,
   0.5321613313265069,
   1.060679733181632,
   0.7964205322540695,
   0.6642909317902882,
   -2.5068194793404635,
   -2.374689878876682,
   0.5321613313265069,
   -1.3176530751664315,
   0.5321613313265069,
   -0.12848667099239972,
   0.2679021303989442,
   -2.5068194793404635,
   -1.3176530751664315,
   0.

In [10]:
import numpy as np

# Creating a zscore column for each numerical column + detecting and removing outliers

In [11]:
for col in stats:
    df['zscore-' + col] = abs(df[col] - stats[col]['mean'])/df[col].std()
    outliers = df[df['zscore-' + col] > 3]
    df = df[df['zscore-' + col] < 3]

In [12]:
for col in df: 
    print(col, len(df[col].unique()))

video_id 6233
trending_date 205
title 6335
channel_title 2182
category_id 15
publish_time 6153
tags 5942
views 38449
likes 27897
dislikes 6961
comment_count 11901
thumbnail_link 6234
comments_disabled 2
ratings_disabled 2
video_error_or_removed 2
description 6740
zscore-category_id 15
zscore-views 38449
zscore-likes 27897
zscore-dislikes 6961
zscore-comment_count 11901


In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38917 entries, 0 to 41413
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   video_id                38917 non-null  object 
 1   trending_date           38917 non-null  object 
 2   title                   38917 non-null  object 
 3   channel_title           38917 non-null  object 
 4   category_id             38917 non-null  float64
 5   publish_time            38917 non-null  object 
 6   tags                    38917 non-null  object 
 7   views                   38917 non-null  float64
 8   likes                   38917 non-null  float64
 9   dislikes                38917 non-null  float64
 10  comment_count           38917 non-null  float64
 11  thumbnail_link          38917 non-null  object 
 12  comments_disabled       38917 non-null  object 
 13  ratings_disabled        38917 non-null  object 
 14  video_error_or_removed  38917 non-null

# we can change the type of comments_disabled	ratings_disabled	video_error_or_removed to int by repalcing true with 1 and false with 0

In [14]:
n_tags = len(df['tags'])

In [15]:
n_tags

38917

In [45]:
df['tags'].mode()

0    [none]
dtype: object